File: worker/src/credits.py
"""Meranie a účtovanie LLM volaní.

MeteredLlm má rovnaké verejné metódy ako Llm — moduly predlohy nič nepoznajú.
Pred volaním kontrola zostatku (≤0 → OutOfCredits + pauza modelu), po volaní
zápis do ledgeru: atlas cena z cenníka, klientovi × multiplier.

Neobmedzené účty (`accounts.unlimited`) kontrolu preskakujú — nikdy sa
nepauzujú a nič sa im neodpočíta, ale do ledgeru sa zapisujú rovnako.

ÚČTUJE SA AJ PÁD
----------------
`Llm._chat` opakuje pokus po 429/5xx a poskytovateľ si tokeny z každého pokusu
účtuje. Preto sa `last_usage` v `llm.py` SČÍTAVA cez celé logické volanie a
`_bill` beží aj na chybovej vetve — inak by trojnásobný pokus, ktorý nakoniec
zlyhal, nezaplatil nikto (okrem nás).
"""
from __future__ import annotations
import logging

log = logging.getLogger(__name__)

# `chat` je vyhradený pre správu, ktorú modelka naozaj napísala a odoslala —
# dashboard z počtu týchto riadkov robí dlaždicu „Replies sent". `structured`
# preto NIE JE chat: to sú pomocné volania na pozadí jednej odpovede (sudca,
# pamäťové tvrdenia, fakty, recall, hovorená forma hlasovky), na Simone ich
# vychádza ~0.4 na odpoveď. Kým sa delili o ten istý štítok, klient videl
# násobok toho, čo modelka reálne odoslala.
KIND_BY_METHOD = {"reply": "chat", "structured": "assist", "summarize": "summary",
                  "describe_image": "vision", "transcribe_voice": "audio"}

# Rovnaká jednotka aj slovník ako v appke (`web/lib/credits.ts`) — majiteľ číta
# obe hlášky a nesmie z nich mať pocit, že ide o dva rôzne zostatky.
OUT_OF_CREDITS_MSG = (
    "You are out of Pipe Coins — she has stopped replying. "
    "Top up your balance and she picks up where she left off."
)

# Chýbajúci cenník je konfiguračná chyba, nie udalosť — warning stačí raz za
# proces a slug, inak by log zaplavila každá jedna odpoveď každého tenanta.
_MISSING_PRICE_WARNED: set = set()


class OutOfCredits(RuntimeError):
    pass


class MeteredLlm:
    def __init__(self, llm, registry, model_id: str, model_slug: str,
                 fallback_per_mtok: float = 5.0, notify=None) -> None:
        self._llm = llm
        self._reg = registry
        self._model_id = model_id
        self._slug = model_slug
        self._fallback = fallback_per_mtok
        # Async callable (typicky ControlBot.notify) — správa majiteľovi, že
        # kredit došiel. None = kontrolný bot nebeží, len sa pauzuje.
        self._notify = notify
        self._paused_notified = False

    def __getattr__(self, name):          # metódy Llm, ktoré nemeriame (close…)
        return getattr(self._llm, name)

    async def _metered(self, method: str, *args, **kwargs):
        # Neobmedzený účet (superadmin, interné modelky) sa nekontroluje vôbec:
        # žiadna pauza, žiadna správa majiteľovi. Meranie tým NEKONČÍ — usage
        # sa zapíše nižšie rovnako a `record_usage` len na strane DB preskočí
        # odpočet, takže reálna spotreba ostáva v admin číslach vidieť.
        balance, unlimited = await self._reg.credit_state(self._model_id)
        if not unlimited and balance <= 0:
            # Kým sa runner zastaví (pool ho odpojí do ~30 s), stihne prísť
            # ďalších správ — stav aj správu majiteľovi preto posielame len
            # pri PRVOM zistení, inak by to bol spam do chatu aj do DB.
            if not self._paused_notified:
                self._paused_notified = True
                await self._reg.set_status(self._model_id, "paused", "out_of_credits")
                if self._notify is not None:
                    try:
                        await self._notify(OUT_OF_CREDITS_MSG)
                    except Exception:  # noqa: BLE001 — nedoručená správa nič nemení
                        log.exception("Správa o vyčerpaných kreditoch neodišla")
            raise OutOfCredits(self._model_id)
        try:
            result = await getattr(self._llm, method)(*args, **kwargs)
        except Exception:
            # Volanie padlo, ale tokeny už zhoreli — `Llm._chat` si spotrebu
            # zo VŠETKÝCH pokusov sčítava do `last_usage`, takže tu je čo
            # zaúčtovať. Bez tohto by neúspešné volanie (typicky tri pokusy
            # po 429) platil Atlas z vlastného. Výnimka ide ďalej nezmenená.
            await self._bill(method)
            raise
        await self._bill(method)
        return result

    async def _bill(self, method: str) -> None:
        """Zapíše spotrebu posledného volania do ledgeru. Nikdy nehádže.

        Volá sa aj po páde volania — preto je celé telo best-effort: účtovanie
        nesmie prekryť pôvodnú chybu ani zhodiť úspešnú odpoveď.
        """
        try:
            await self._bill_inner(method)
        except Exception:  # noqa: BLE001 — účtovanie nikdy nezhodí volanie
            log.exception("Zaúčtovanie spotreby zlyhalo")

    async def _bill_inner(self, method: str) -> None:
        usage = getattr(self._llm, "last_usage", None) or {}
        i, o = int(usage.get("input", 0) or 0), int(usage.get("output", 0) or 0)
        if i <= 0 and o <= 0:
            # Volanie nedošlo k modelu (DNS, timeout pred odpoveďou) — zapisovať
            # nulový riadok by len zaplnil ledger bez informácie.
            return
        price = await self._reg.pricing(self._slug)
        if price.get("input_usd_per_mtok") or price.get("output_usd_per_mtok"):
            atlas = i / 1e6 * float(price["input_usd_per_mtok"]) \
                  + o / 1e6 * float(price["output_usd_per_mtok"])
        else:
            if self._slug in _MISSING_PRICE_WARNED:
                log.debug("Chýba cenník pre %s — fallback %.2f/Mtok", self._slug, self._fallback)
            else:
                _MISSING_PRICE_WARNED.add(self._slug)
                log.warning("Chýba cenník pre %s — fallback %.2f/Mtok", self._slug, self._fallback)
            atlas = (i + o) / 1e6 * self._fallback
        charged = atlas * float(price.get("multiplier", 2.0))
        try:
            await self._reg.record_usage(self._model_id, KIND_BY_METHOD[method],
                                         i, o, 0, round(atlas, 6), round(charged, 6))
        except Exception:
            log.exception("Zápis usage zlyhal — odpoveď ide ďalej, dorovná sa ďalším volaním")

    async def charge_unit(self, kind: str, key: str, default_price: float) -> None:
        """Zauctuje jeden kus (hlasovka/fotka) po USPESNOM odoslani.

        Je to tu, a nie v userbote, aby vsetko uctovanie islo cez jedno miesto —
        tokeny aj kusy. Cena sa cita z `app_config` v DB (tu istu hodnotu
        zobrazuje web), nie z konstanty v kode.
        """
        await self._reg.charge_unit(self._model_id, kind, key, default_price)

    async def reply(self, *a, **kw): return await self._metered("reply", *a, **kw)
    # `suggest` sem pribudlo neskôr než ostatné a chvíľu chýbalo — padalo cez
    # `__getattr__` rovno na `Llm`, takže KAŽDÝ návrh v poloautomate, každé
    # pregenerovanie a každá veta z generátora sa modelu zaplatili, ale
    # klientovi sa nezaúčtovali. Meraný musí byť každý spôsob, ako sa dá
    # minúť token.
    async def suggest(self, *a, **kw): return await self._metered("suggest", *a, **kw)
    async def structured(self, *a, **kw): return await self._metered("structured", *a, **kw)
    async def summarize(self, *a, **kw): return await self._metered("summarize", *a, **kw)
    async def describe_image(self, *a, **kw): return await self._metered("describe_image", *a, **kw)
    async def transcribe_voice(self, *a, **kw): return await self._metered("transcribe_voice", *a, **kw)

File: web/lib/credits.ts
/**
 * Kredit a účtovanie LLM volaní z webu.
 *
 * Toto je webová dvojička `worker/src/credits.py` a matematika musí sedieť na
 * cent: cena z tabuľky `pricing` za milión tokenov, klientovi × `multiplier`,
 * chýbajúci cenník → fallback za Mtok. Keby web počítal inak než worker, ten
 * istý model by stál dve rôzne sumy podľa toho, kto ho zavolal.
 *
 * SERVER ONLY — `record_usage` beží cez service key (rola `authenticated` na
 * `pricing` ani na `usage_events` grant nemá a mať nebude: zápis do ledgeru je
 * vec servera, nie prehliadača).
 */

import { createClient, createServiceClient } from "@/lib/supabase/server";
import { toNumber } from "@/lib/format";
import {
  DEFAULT_MULTIPLIER,
  FALLBACK_PER_MTOK,
  computeCost,
  type PriceRow,
  type UsageCost,
} from "@/lib/pricing";

export type { PriceRow, UsageCost } from "@/lib/pricing";

export type CreditState = { balance: number; unlimited: boolean };

/**
 * Zostatok prihláseného účtu. `credit_state` RPC je len pre service_role
 * (migrácia 013), ale `authenticated` má select na `accounts` aj na stĺpec
 * `unlimited` — a RLS vráti výhradne vlastný riadok, takže user-scoped čítanie
 * je tu presnejšie než obchádzanie RLS service kľúčom.
 */
export async function creditState(): Promise<CreditState> {
  const supabase = await createClient();
  const { data } = await supabase
    .from("accounts")
    .select("credit_balance_usd, unlimited")
    .maybeSingle();

  return {
    balance: toNumber(data?.credit_balance_usd as string | number | undefined),
    unlimited: Boolean(data?.unlimited),
  };
}

/**
 * Má účet na čo míňať? Pravidlo je to isté ako v `MeteredLlm._metered`:
 * neobmedzený účet sa nekontroluje, ostatným musí zostatok ostať kladný.
 */
export function hasCredit(state: CreditState): boolean {
  return state.unlimited || state.balance > 0;
}

export const OUT_OF_CREDITS_MSG =
  "You are out of credits. Top up your balance and the AI helper will work again.";

/**
 * Zápis spotreby do ledgeru. Nikdy nehádže — účtovanie nesmie zhodiť odpoveď,
 * ktorú klient už dostal (rovnaké pravidlo ako `MeteredLlm._bill`).
 */
export async function recordUsage(
  modelId: string,
  kind: "chat" | "summary" | "vision" | "audio" | "voice",
  slug: string,
  usage: { input: number; output: number },
): Promise<UsageCost | null> {
  const input = Math.max(0, Math.round(usage.input));
  const output = Math.max(0, Math.round(usage.output));
  // Volanie sa k modelu nedostalo (DNS, timeout pred odpoveďou) — nulový riadok
  // by ledger len zaplnil bez informácie.
  if (input <= 0 && output <= 0) return null;

  try {
    const supabase = createServiceClient();
    const price = await pricing(slug);
    const cost = computeCost(input, output, price, fallbackPerMtok());

    const { error } = await supabase.rpc("record_usage", {
      p_model: modelId,
      p_kind: kind,
      p_input_tokens: input,
      p_output_tokens: output,
      p_unit_count: 0,
      p_atlas_cost_usd: cost.atlas,
      p_charged_usd: cost.charged,
    });
    if (error) {
      console.error("record_usage failed", error.message);
      return null;
    }
    return cost;
  } catch (error) {
    console.error("record_usage failed", error);
    return null;
  }
}

/** Cenník pre slug, inak riadok `_default` — presne ako `Registry.pricing`. */
async function pricing(slug: string): Promise<PriceRow> {
  const supabase = createServiceClient();
  const { data } = await supabase
    .from("pricing")
    .select("model_slug, input_usd_per_mtok, output_usd_per_mtok, multiplier")
    .in("model_slug", [slug, "_default"]);

  const rows = data ?? [];
  const row = rows.find((item) => item.model_slug === slug) ?? rows.find((item) => item.model_slug === "_default");

  return {
    input_usd_per_mtok: toNumber(row?.input_usd_per_mtok as string | number | undefined),
    output_usd_per_mtok: toNumber(row?.output_usd_per_mtok as string | number | undefined),
    multiplier: row?.multiplier === undefined ? DEFAULT_MULTIPLIER : toNumber(row.multiplier),
  };
}

/** Prepísateľná fallback cena — worker číta tú istú premennú. */
function fallbackPerMtok(): number {
  const raw = Number(process.env.FALLBACK_PRICE_PER_MTOK);
  return Number.isFinite(raw) && raw > 0 ? raw : FALLBACK_PER_MTOK;
}
